RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,332 @@
+`super().__torch_function__` would call `SubSubTensor.__add__` (realise, this
+is a different object than `Tensor.__add__`), which, not realising that it has
+been processed already, would pass control *back* to
+`SubTensor.__torch_function__` and so on ad infinitum.
+
+In this instance, with the proposed changes, processing would follow the
+`__torch_function__` protocol. This means that control would end up in
+`SubTensor.__torch_function__`, go to `Tensor.__torch_function__` from there
+and and then come to `SubSubTensor.__add__`, from where it would go to
+`Tensor.__add__`, and then back up the stack in the reverse order. This means
+that great care needs to be taken when writing `SubTensor.__torch_function__`
+to take into account the fact that it has to handle subclass methods.
+
+### Protocol support for external libraries
+We will also recommend that all `Tensor` subclasses make their own methods go
+through `__torch_function__` via a decorator `@torch_function_dispatch`. This
+decorator was added and then removed for performance reasons, however it will
+be added back to allow external libraries to interface with the protocol. It
+will take a single argument: a dispatcher, i.e. a callable that returns an
+iterable of all the "duck-Tensors", or possible candidates for classes that may
+implement `__torch_function__`.
+
+If a library forgets to add the aforementioned decorator, then the method will
+no longer dispatch at all to any form of `__torch_function__`. In other words,
+it will lose support for the protocol. This can lead to confusion, as some
+methods of the subclass will pass through `__torch_function__` (the ones
+inherited from `torch.Tensor`), and some won't.
+
+Note that subclasses will still be passed through due to the default
+implementation of `__torch_function__`, but any `__torch_function__` defined on
+the class itself (or any of its subclasses) won't have an effect on its
+methods.
+
+This is a design choice that a subclass author will have to make, whether they
+prefer their own functions/methods to pass through `__torch_function__` like
+PyTorch's implementations, or whether they'd like ultimately to not support the
+protocol and accept having a mix of overridable and non-overridable methods.
+
+We do not propose automatic marking of functions with this decorator due to the
+potential backwards-compatibility break it could cause, as well as the
+parameters that are needed in order to allow this to happen (namely the
+dispatcher, which isn't in our control).
+
+### Making `torch.Tensor._make_subclass` public API
+`torch.Tensor._make_subclass` will be renamed to `torch.Tensor.make_subclass`
+and it will become public API. This method will create an object that has the
+same data pointer as the original object, which means that modifications to
+this will be reflected in the original object. More or less, it will have the
+same effect as modifying an object's `__class__` attribute in Python.
+
```

Comment:
This still needs a motivation, it's not explained anywhere I think. Is it _only_ for use within `SubTensor.__torch_function__` as in the one example, or for other reasons too? Please give the 1-2 sentence motivation, and perhaps a concrete example of how to use it if there's another reason.

Reply:
Added just below this comment.